src/opm/input/eclipse/EclipseState/Grid/EclipseGrid.cpp

```diff
@@ -332,22 +333,65 @@ EclipseGrid::EclipseGrid(const Deck& deck, const int * actnum)
         return this->m_circle;
     }
 
+
     void EclipseGrid::initGrid(const Deck& deck, const int* actnum)
     {
-        if (deck.hasKeyword<ParserKeywords::RADIAL>()) {
+        enum GridType { COORD, DEPTHZ, TOPS, RADIAL, SPIDER, GDFILE, GT_SIZE };
+
+        std::vector<int> found;
+        if (hasCornerPointKeywords(deck)) {
+            found.push_back(COORD);
+        }
+        if (hasDVDEPTHZKeywords(deck)) {
+            found.push_back(DEPTHZ);
+        }
+        if (hasDTOPSKeywords(deck)) {
+            found.push_back(TOPS);
+        }
+        if (hasRadialKeywords(deck)) {
+            found.push_back(RADIAL);
+        }
+        if (hasSpiderKeywords(deck)) {
+            found.push_back(SPIDER);
+        }
+        if (hasGDFILE(deck)) {
+            found.push_back(GDFILE);
+        }
+
+        const std::string messages[GT_SIZE] {
+            "COORD with ZCORN creates a corner-point grid",
+            "DEPTHZ with DXV, DYV, DZV creates a cartesian grid",
+            "TOPS with DX/DXV, DY/DYV, DZ/DZV creates a cartesian grid",
+            "RADIAL with DR/DRV, DTHETA/DTHETAV, DZ/DZV and TOPS creates a cylindrical grid",
+            "SPIDER with DR/DRV, DTHETA/DTHETAV, DZ/DZV and TOPS creates a spider grid",
+            "GDFILE reads a grid from file"};
+
+        if (found.size() == 0) {
+            std::string message = "The grid must be specified using one of these options:";
+            for (int grid_type = 0; grid_type < GT_SIZE; grid_type++) {
+                message += "\n    " + messages[grid_type];
+            }
+            throw std::invalid_argument(message);
+        }
+
+        if (found.size() > 1) {
+            std::string message = "The specification of the grid is ambiguous:";
+            for (const int& grid_type : found) {
+                message += "\n    " + messages[grid_type];
+            }
+            throw std::invalid_argument(message);
+        }
+
+        if (found[0] == COORD) {
+            initCornerPointGrid(deck);
+        } else if (found[0] == DEPTHZ || found[0] == TOPS) {
+            initCartesianGrid(deck);
+        } else if (found[0] == RADIAL) {
             initCylindricalGrid(deck );
-        } else if (deck.hasKeyword<ParserKeywords::SPIDER>()) {
+        } else if (found[0] == SPIDER) {
             initSpiderwebGrid(deck );
-        } else {
-            if (hasCornerPointKeywords(deck)) {
-                initCornerPointGrid(deck);
-            } else if (hasCartesianKeywords(deck)) {
-                initCartesianGrid(deck);
-            } else if (hasGDFILE(deck)) {
-                initBinaryGrid(deck);
-            } else {
-                throw std::invalid_argument("EclipseGrid needs cornerpoint or cartesian keywords.");
-            }
+        } else if (found[0] == GDFILE) {
+            initBinaryGrid(deck);
         }
```

Comment:
I think I'd prefer that this be written as a `switch`/`case` instead:
```C++
switch (found.front()) {
case GridType::COORD:
    this->initCornerPointGrid(deck);
    break;

case GridType::DEPTHZ:
case GridType::TOPS:
    this->initCartesianGrid(deck);
    break;

case GridType::RADIAL:
    this->initCylindricalGrid(deck);
    break;

case GridType::SPIDER:
    this->initSpiderWebGrid(deck);
    break;

case GridType::GDFILE:
    this->initBinaryGrid(deck);
    break;
}
```
since that's easier to generalise later.

Reply:
Done.